src/main/kotlin/me/peckb/aoc/_2024/calendar/day09/Day09.kt

```diff
@@ -0,0 +1,140 @@
+        file = false
+      } else {
+        memory.add(Space.Empty(length))
+        file = true
+      }
+    }
+
+    return memory
+  }
+
+  private fun defrag(memory: MutableList<Space>, splitFiles: Boolean) {
+    var lastFileIndex = (memory.lastIndex downTo 0).first { memory[it] is Space.Full }
+    var spaceIndex = 1
+
+    while(spaceIndex < memory.size && (spaceIndex < lastFileIndex)) {
+      val lastFile = (lastFileIndex downTo spaceIndex)
+        .first { i -> memory[i] is Space.Full }
+        .also { lastFileIndex = it }
+        .let { memory[it] as Space.Full }
+
+      val emptySpace = (spaceIndex until lastFileIndex)
+        .first { i -> memory[i] is Space.Empty }
+        .also { spaceIndex = it }
+        .let { memory[it] as Space.Empty }
+
+      val availableSpace = emptySpace.length
+      val fileSize = lastFile.length
+
+      if (fileSize < availableSpace) {
+        emptySpace.length = (availableSpace - fileSize)
+        memory.add(spaceIndex, Space.Full(fileSize, lastFile.file))
+        spaceIndex++
+        memory.removeAt(lastFileIndex + 1)
+
+      } else if (fileSize == availableSpace) {
+        emptySpace.length = 0
+        memory.add(spaceIndex, Space.Full(availableSpace, lastFile.file))
+        memory.removeAt(lastFileIndex-- + 1)
+        memory.removeAt(spaceIndex + 1)
+
+      } else { // fileSize > availableSpace
+        if (splitFiles) {
+          emptySpace.length = 0
+          memory.add(spaceIndex, Space.Full(availableSpace, lastFile.file))
+          memory.removeAt(spaceIndex + 1)
+
+          lastFile.length -= availableSpace
+
+        } else {
+          var tempSpaceIndex = spaceIndex + 1
+          var nextEmptySpace = (spaceIndex until lastFileIndex).firstOrNull { i ->
+            memory[i] is Space.Empty
+          }?.let { tempSpaceIndex = it; memory[it] as Space.Empty }
+
+          while (nextEmptySpace != null && tempSpaceIndex < lastFileIndex) {
+            if (fileSize <= nextEmptySpace.length) {
+              if (fileSize < nextEmptySpace.length) {
+                nextEmptySpace.length = (nextEmptySpace.length - fileSize)
+                memory.add(tempSpaceIndex, Space.Full(fileSize, lastFile.file))
+                memory[++lastFileIndex] = Space.Empty(fileSize)
+
+              } else { // amountNeededToTake = nextEmptySpace.length
+                nextEmptySpace.length = 0
+                memory.add(tempSpaceIndex, Space.Full(fileSize, lastFile.file))
+                memory[++lastFileIndex] = Space.Empty(fileSize)
+              }
+
+              nextEmptySpace = null
+            } else {
+              nextEmptySpace = (tempSpaceIndex + 1 until lastFileIndex).firstOrNull { i ->
+                memory[i] is Space.Empty
+              }?.let { tempSpaceIndex = it; memory[it] as Space.Empty }
+            }
+          }
+
+          lastFileIndex--
+        }
+      }
+    }
+  }
+
+  private fun checksum(memory: MutableList<Space>): Long {
```

Comment:
mapindexed(I, x) with a reduce / fold somewhere?

Reply:
The problem is that the index on the list would count the item in the list, rather than counting what "index" it would be in memory

Reply:
your memory isn't the "final list"?

Reply:
oh... I get it. you have 
[(full len 5),( space Len 1)] so that would be index 0 and 1. not index 0-4, 5 in the list